### PR TITLE
Define default Python engine by system and user

### PR DESCRIPTION
**Pull request by mmisol**
### Purpose

Adds the ability to define a default Python engine for the system. This
is done through a new property in the model start configuration, named
'DefaultPythonEngine'.

Also adds the ability for users to define through a setting their own
default. This can be set to either engine or to fallback to the system
default, whatever it is.

Serialization is tweaked so that 'IronPython2' remains the
deserialization default but is not the absolute default. Engine will be
unspecified on node creation but become loaded as soon as the property
is accessed, to avoid errors on automatic execution.

In order to make system and user settings available on the node, and
avoiding further changes to the API, the properties were made static
and internal.

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [x] This PR modifies some build requirements and the readme is updated

### Reviewers

@mjkkirschner @QilongTang 

### FYIs

@DynamoDS/dynamo

### CODE REVIEW

**Comment by mmisol** [comment]
Just wanted to make it clear this is a Work In Progress. Please ignore the commented out code. The idea is to get some feedback on how this looks before wrapping up, since there are several questionable decisions made.

File: src/DynamoApplications/StartupUtils.cs [hunk 1]
```diff
@@ -109,8 +111,9 @@ public static CommandLineArguments Parse(string[] args)
...
                 .Add("i=|I=|import", "Import, Instruct Dynamo to import an assembly as a node library. This argument should be a filepath to a single .dll" +
-                " - if you wish to import multiple dlls - use this flag multiple times: -i 'assembly1.dll' -i 'assembly2.dll' ", i => importPaths.Add(i)).
-                Add("gp=|GP=|geometrypath=|GeometryPath=", "relative or absolute path to a directory containing ASM. When supplied, instead of searching the hard disk for ASM, it will be loaded directly from this path.", gp => asmPath = gp);
+                " - if you wish to import multiple dlls - use this flag multiple times: -i 'assembly1.dll' -i 'assembly2.dll' ", i => importPaths.Add(i))
+                .Add("gp=|GP=|geometrypath=|GeometryPath=", "relative or absolute path to a directory containing ASM. When supplied, instead of searching the hard disk for ASM, it will be loaded directly from this path.", gp => asmPath = gp);
+                //.Add("py=|PY=|pythonEngine=|PythonEngine=", "Default Python script engine", py => pythonEngine = py);
```
Comment (mmisol) [inline src/DynamoApplications/StartupUtils.cs:116]: At some point I thought this could also be a command line option, but then disregarded that. Integrators won't use that and not sure we would ever use it for Sandbox. It's also not fully implemented.

File: src/DynamoCore/Graph/Nodes/NodeModel.cs [hunk 2]
```diff
@@ -2515,6 +2515,11 @@ public static NodeMigrationData MigrateShortestLacingToAutoLacing(NodeMigrationD
 
         #endregion
 
+        public virtual void OnBeforeAdd(Dynamo.Models.DynamoModel dynamoModel, WorkspaceModel workspace)
+        {
+            // Purposely empty. For inheritors to override if needed.
+        }
```
Comment (mmisol) [inline src/DynamoCore/Graph/Nodes/NodeModel.cs:2521]: In my mind it made sense to add this sort of life-cycle method to the node. It allowed to act at the right time and provides the required `DynamoModel`.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 3]
```diff
@@ -501,6 +511,14 @@ public interface IStartConfiguration
...
             /// </summary>
             bool IsHeadless { get; set; }
+            /// <summary>
+            /// Default Python script engine
+            /// </summary>
+            string DefaultPythonEngine { get; set; }
+            /// <summary>
+            /// Provide values for startup options
+            /// </summary>
+            //IDictionary<string, string> StartupOptions { get; set; }
```
Comment (mmisol) [inline src/DynamoCore/Models/DynamoModel.cs:521]: This is like an alternative API I had in mind at first. It would allow to provide command-line options in a generic way. The only benefit I can think about it is that we don't add an explicit Python property to the config. On the other hand, it is messier to add entries to a dictionary than assigning properties.

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 4]
```diff
@@ -1108,6 +1108,48 @@ public object ConvertBack(object value, Type targetType, object parameter, Syste
...
+    //        return DependencyProperty.UnsetValue;
+    //    }
+
+    //    public object[] ConvertBack(object value, Type[] targetTypes, object parameter, CultureInfo culture)
+    //    {
+    //        // No way to implement this. Multi-binding can't achieve this
+    //        throw new NotImplementedException();
+    //    }
+    //}
+
```
Comment (mmisol) [inline src/DynamoCoreWpf/UI/Converters.cs:1152]: Some experiments I did with WPF before ending up with the current UI.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml [hunk 5]
```diff
@@ -724,6 +724,55 @@
                                   IsCheckable="True"
                                   IsChecked="{Binding ShowTabsAndSpacesInScriptEditor, Mode=TwoWay}"
                                   Header="{x:Static p:Resources.DynamoViewSettingsMenuShowTabsAndSpacesInScriptEditor}"/>
+                        <MenuItem Focusable="False"
+                                  Header="{x:Static p:Resources.DynamoViewSettingsDefaultPythonEngine}">
+                            <ComboBox ItemsSource="{Binding DefaultPythonEngineOptions}" SelectedValue="{Binding DefaultPythonEngine, Mode=TwoWay}" SelectedValuePath="Value" DisplayMemberPath="Name"/>
+                        </MenuItem>
```
Comment (mmisol) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml:730]: Used a drop down. As you can see from the commented markup I also tried child checkable menu items, which had some limitations. Open to discussion.

File: src/Libraries/PythonNodeModels/PythonNode.cs [hunk 6]
```diff
@@ -29,13 +32,15 @@ public PythonCodeMigrationEventArgs(string oldCode, string newCode)
 
     public abstract class PythonNodeBase : VariableInputNode
     {
-        private PythonEngineVersion engine = PythonEngineVersion.IronPython2;
+        private PythonEngineVersion? engine = null;
```
Comment (mmisol) [inline src/Libraries/PythonNodeModels/PythonNode.cs:35]: The engine became nullable, as it cannot be initialized at this point. We have no access to `DynamoModel`. Luckily JSON serialization handles this change without problems.

File: src/Libraries/PythonNodeModels/PythonNode.cs [hunk 7]
```diff
@@ -54,6 +59,30 @@ protected PythonNodeBase()
...
+                }
+                else if (!string.IsNullOrEmpty(systemDefault) && Enum.TryParse(setting, out version))
+                {
+                    Engine = version;
+                }
+                else
+                {
+                    // In the absence of both a setting and system default, default to deserialization default.
+                    Engine = PythonEngineVersion.IronPython2;
+                }
```
Comment (mmisol) [inline src/Libraries/PythonNodeModels/PythonNode.cs:82]: user setting > system default > deserialization default

File: src/DynamoCore/Configuration/PreferenceSettings.cs [hunk 8]
```diff
@@ -333,6 +333,11 @@ public string PythonTemplateFilePath
         /// </summary>
         public bool ShowTabsAndSpacesInScriptEditor { get; set; }
 
+        /// <summary>
+        /// Engine used by default for new Python script and string nodes. If defined this takes precedence over any system settings.
```
Comment (mjkkirschner) [inline src/DynamoCore/Configuration/PreferenceSettings.cs:337]: do you mean if not null? Does empty string qualify as defined?

File: src/DynamoCore/Graph/Nodes/NodeModel.cs [hunk 9]
```diff
@@ -2515,6 +2515,11 @@ public static NodeMigrationData MigrateShortestLacingToAutoLacing(NodeMigrationD
 
         #endregion
 
+        public virtual void OnBeforeAdd(Dynamo.Models.DynamoModel dynamoModel, WorkspaceModel workspace)
```
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Nodes/NodeModel.cs:2518]: this is quite a big change - in the rest of Dynamo it's common for models to only have references to their children - ie DynamoModel knows about NodeModels - and NodeModels know about portModels, but not the other way around.

Curious why you need this reference - maybe it will become clear.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 10]
```diff
@@ -501,6 +511,14 @@ public interface IStartConfiguration
             /// No update checks or analytics collection should be done.
             /// </summary>
             bool IsHeadless { get; set; }
+            /// <summary>
+            /// Default Python script engine
+            /// </summary>
+            string DefaultPythonEngine { get; set; }
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs:517]: modifying this interface is an API break - we can add a new interface instead, and merge them together in 3.x

File: src/DynamoCore/Models/DynamoModel.cs (hunk 3 above)
Comment (mjkkirschner) [reply src/DynamoCore/Models/DynamoModel.cs:521]: I've always thought it would be nice to be able to override the entire config with an xml or json file - but thats a discussion for another time I think.

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 11]
```diff
@@ -46,6 +47,20 @@ public interface IDynamoViewModel : INotifyPropertyChanged
         ObservableCollection<WorkspaceViewModel> Workspaces { get; set; } 
     }
 
+    public class SimpleViewModel<T> : ViewModelBase, INotifyPropertyChanged
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:50]: experiment?

File: src/Libraries/PythonNodeModels/PythonNode.cs [hunk 12]
```diff
@@ -29,13 +32,15 @@ public PythonCodeMigrationEventArgs(string oldCode, string newCode)
...
+        private PythonEngineVersion? engine = null;
 
         [JsonConverter(typeof(StringEnumConverter))]
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Populate)]
+        [DefaultValue("IronPython2")]
         /// <summary>
         /// Return the user selected python engine enum.
         /// </summary>
-        public PythonEngineVersion Engine
+        public PythonEngineVersion? Engine
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModels/PythonNode.cs:43]: I wonder if this qualifies as an API break...

File: src/Libraries/PythonNodeModels/PythonNode.cs [hunk 13]
```diff
@@ -88,7 +117,7 @@ protected override string GetInputTooltip(int index)
 
             // Here we switched to use the AstFactory.BuildFunctionCall version that accept
             // class name and function name. They will be set by PythonEngineSelector by the engine value. 
-            PythonEngineSelector.Instance.GetEvaluatorInfo(Engine, out string evaluatorClass, out string evaluationMethod);
+            PythonEngineSelector.Instance.GetEvaluatorInfo(Engine.Value, out string evaluatorClass, out string evaluationMethod);
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModels/PythonNode.cs:120]: I guess so?

File: src/DynamoCore/Graph/Nodes/NodeModel.cs (hunk 9 above)
Comment (mjkkirschner) [reply src/DynamoCore/Graph/Nodes/NodeModel.cs:2518]: hmm - it makes a lot of sense that the python nodeModel needs to continue working even in headless Dynamo.

File: src/Libraries/PythonNodeModels/PythonNode.cs [hunk 14]
```diff
@@ -54,6 +59,30 @@ protected PythonNodeBase()
             ArgumentLacing = LacingStrategy.Disabled;
         }
 
+        public override void OnBeforeAdd(DynamoModel dynamoModel, WorkspaceModel workspace)
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModels/PythonNode.cs:62]: workspace looks unused - just thinking ahead?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 11 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:50]: apparently you can use old style tuple items as bindings - you cannot use new style value tuples though as those members are fields - anyhow, something I consider I guess.

Comment (mjkkirschner) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:50]: https://stackoverflow.com/questions/43208852/is-it-possible-to-bind-to-a-valuetuple-field-in-wpf-with-c7

File: src/DynamoCore/Graph/Nodes/NodeModel.cs (hunk 9 above)
Comment (mjkkirschner) [reply src/DynamoCore/Graph/Nodes/NodeModel.cs:2518]: I'm curious what you think of using the active `ExecutionSession` instead - storing the pythonEngine there and extracting it from the static map when the node is constructed.

File: src/DynamoCore/Configuration/PreferenceSettings.cs (hunk 8 above)
Comment (mmisol) [reply src/DynamoCore/Configuration/PreferenceSettings.cs:337]: I'm pretty sure I'm using empty string for not defined.

File: src/DynamoCore/Graph/Nodes/NodeModel.cs (hunk 9 above)
Comment (mmisol) [reply src/DynamoCore/Graph/Nodes/NodeModel.cs:2518]: I'm open to change this. `ExecutingSession` is looking uglier each time we add another new thing to it, but I guess it can be used.

This does expose more data, but we are not being super consistent with the visibility anyway. For view models the rules are pretty different than what you mention here.

I don't think this would break headless Dynamo, unless you added a node to a graph dynamically (can you do that headless?) and forgot to assign an `Engine` to it.

File: src/DynamoCore/Models/DynamoModel.cs (hunk 10 above)
Comment (mmisol) [reply src/DynamoCore/Models/DynamoModel.cs:517]: An alternative can be not have this in the interface but keep it in `DefaultStartConfiguration`, and cast and see in `DynamoModel` constructor. I see some other properties on this boat already.

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 11 above)
Comment (mmisol) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:50]: Maybe not the right place for it, but I thought this could be a good way to to it. I'm not a fan of tuples, specially because the property names are positional and have no meaning.

File: src/Libraries/PythonNodeModels/PythonNode.cs (hunk 12 above)
Comment (mmisol) [reply src/Libraries/PythonNodeModels/PythonNode.cs:43]: It does.

File: src/Libraries/PythonNodeModels/PythonNode.cs (hunk 14 above)
Comment (mmisol) [reply src/Libraries/PythonNodeModels/PythonNode.cs:62]: In my mind, this "life-cycle method" made more sense if you only got the workspace where the node was being added. However, you have the same problem, `WorkspaceModel` has no visibility of the `DynamoModel` associated with it. In the end I kept both because only providing the `DynamoModel` made less sense as an API, but it's true it's not really used.

File: src/DynamoCore/Configuration/PreferenceSettings.cs (hunk 8 above)
Comment (mjkkirschner) [reply src/DynamoCore/Configuration/PreferenceSettings.cs:337]: can you document that in this comment?

File: src/DynamoCore/Graph/Nodes/NodeModel.cs (hunk 9 above)
Comment (mjkkirschner) [reply src/DynamoCore/Graph/Nodes/NodeModel.cs:2518]: you can add a node to a graph dynamically in headless mode for sure - consider some headless client of Dynamo that calls the createNodeCommand - but - I was actually trying to make the point that I thought your solution was better than my original suggestion because this doesn't require the NodeView to be around to correctly set the engine.

Comment (mjkkirschner) [reply src/DynamoCore/Graph/Nodes/NodeModel.cs:2518]: for some context - I believe historically, there was a giant static class called something like `DynamoSettings` or `DynamoModelEntryPoint` that was used from everywhere, this made the separation between core and wpf layers very difficult to enforce, and a big refactor was undertaken.

There was also the idea that one day we'd have multiple DynamoModels running at the same time in the same process, but this hit other roadblocks.

I'm not sure the pain for nodeModel developers (like you here) is worth the old ideas from that refactor any longer.

File: src/Libraries/PythonNodeModels/PythonNode.cs (hunk 12 above)
Comment (mjkkirschner) [reply src/Libraries/PythonNodeModels/PythonNode.cs:43]: that seems like a problem ;) Can we add a new enum like none?

File: src/DynamoCore/Graph/Nodes/NodeModel.cs (hunk 9 above)
Comment (mmisol) [reply src/DynamoCore/Graph/Nodes/NodeModel.cs:2518]: Multiple DynamoModel sounds intriguing. In any case a node would belong to just one of those, right? 

Giving a node model access to its parent workspace model and that one having access to its parent dynamo model does not sound so bad to me. It feels having access to the context a node model is in should be more direct than relying on global classes. 

In any case, we probably don't want to make a refactor part of this PR. I'll stick to `ExecutingSession` for now.

File: src/Libraries/PythonNodeModels/PythonNode.cs (hunk 12 above)
Comment (mmisol) [reply src/Libraries/PythonNodeModels/PythonNode.cs:43]: I can do this. I just felt we weren't tied to this API yet since it was recently added, but I guess people may be using it already.

File: src/DynamoCore/Configuration/PreferenceSettings.cs [hunk 15]
```diff
@@ -388,6 +393,7 @@ public PreferenceSettings()
             PythonTemplateFilePath = "";
             IsIronPythonDialogDisabled = false;
             ShowTabsAndSpacesInScriptEditor = false;
+            DefaultPythonEngine = string.Empty;
```
Comment (QilongTang) [inline src/DynamoCore/Configuration/PreferenceSettings.cs:396]: Do we want to make PythonEngineVersion enum static so we can refer anywhere with known value options?

File: src/DynamoCore/Graph/Nodes/NodeModel.cs (hunk 9 above)
Comment (mmisol) [reply src/DynamoCore/Graph/Nodes/NodeModel.cs:2518]: Now thinking again from a security perspective, accessing the `DynamoModel` opens the door for a node to do a bunch of things we may not want them to do. Note that the risk currently exists for a custom `NodeView` as they can access the `DynamoViewModel` and then to the `DynamoModel`.

It might be a  good idea to think this further and consider refactoring in 3.0.

File: src/DynamoCore/Configuration/PreferenceSettings.cs (hunk 15 above)
Comment (mmisol) [reply src/DynamoCore/Configuration/PreferenceSettings.cs:396]: Do you mean make the `DefaultPythonEngine` preference setting static? I guess that could work, but we also have the system setting, so we'd have to make them both static I guess. It could be an alternative to putting stuff into `ExecutingSession`, which may be cleaner if we can use `InternalsVisibleTo`. Will check that out.

**Comment by mmisol** [comment]
Based on the feedback, I went with internal static properties and got rid of `OnBeforeAdd`. Also cleaned up and added tests. Please take another look @mjkkirschner @QilongTang

File: src/DynamoCore/Models/DynamoModel.cs (hunk 3 above)
Comment (Dewb) [reply src/DynamoCore/Models/DynamoModel.cs]: We already have multiple, partially overlapping systems to manage config, with StartConfiguration, IPreferences, StartupUtils.CommandLineArguments, properties copied into models, etc. Rationalizing all of this should be a priority for the next API-breaking release, would love someone to research and draft a proposal.

**Comment by mmisol** [comment]
Tests are passing now https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/1284/

File: src/DynamoCore/Models/DynamoModel.cs [hunk 16]
```diff
@@ -423,6 +425,14 @@ public IEnumerable<WorkspaceModel> Workspaces
         /// </summary>
         public AuthenticationManager AuthenticationManager { get; set; }
 
+        internal static string DefaultPythonEngine
+        {
+            get
+            {
+                return defaultPythonEngine;
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs:432]: just curious, why is this not using an auto property?

File: src/DynamoCore/Models/DynamoModel.cs [hunk 17]
```diff
@@ -553,6 +567,13 @@ public static DynamoModel Start(IStartConfiguration configuration)
         /// <param name="config">Start configuration</param>
         protected DynamoModel(IStartConfiguration config)
         {
+            if (config is DefaultStartConfiguration)
+            {
+                // This is not exposed in IStartConfiguration to avoid a breaking change.
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs:572]: do you want to add a TODO?

File: src/DynamoCore/Models/DynamoModel.cs [hunk 18]
```diff
@@ -553,6 +567,13 @@ public static DynamoModel Start(IStartConfiguration configuration)
...
+                // This fact should probably be revisited in next major.
+                defaultPythonEngine = ((DefaultStartConfiguration)config).DefaultPythonEngine;
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs:574]: would it ever happen that we've loaded an old xml config and this will throw?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 19]
```diff
@@ -67,7 +68,13 @@ public int ScaleFactorLog
         private readonly DynamoModel model;
         private Point transformOrigin;
         private bool showStartPage = false;
-        
+
+        private ObservableCollection<SimpleViewModel<string>> defaultPythonEngineOptions = new ObservableCollection<SimpleViewModel<string>>();
+        public ObservableCollection<SimpleViewModel<string>> DefaultPythonEngineOptions
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:73]: why the need for the field?

File: src/DynamoCoreWpf/ViewModels/Core/SimpleViewModel.cs [hunk 20]
```diff
@@ -0,0 +1,31 @@
+﻿using System.ComponentModel;
+
+namespace Dynamo.ViewModels
+{
+    /// <summary>
+    /// Dead simple view model for when you only need to have a different display name and underlying value.
+    /// </summary>
+    /// <typeparam name="T">The type of your value</typeparam>
+    public class SimpleViewModel<T> : ViewModelBase, INotifyPropertyChanged
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/SimpleViewModel.cs:9]: do you expect this to ever grow to contain more properties?

Maybe we can call it `NameValuePairViewModel`?

File: src/DynamoCoreWpf/ViewModels/ViewModelBase.cs [hunk 21]
```diff
@@ -20,15 +20,6 @@ public bool IsDebugBuild
             }
         }
 
-        [JsonIgnore]
-        public bool IsPython2ObsoleteDebugModeEnabled
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/ViewModelBase.cs:24]: why is this removed?

File: src/Libraries/PythonNodeModels/PythonNode.cs [hunk 22]
```diff
@@ -29,15 +32,25 @@ public PythonCodeMigrationEventArgs(string oldCode, string newCode)
 
     public abstract class PythonNodeBase : VariableInputNode
     {
-        private PythonEngineVersion engine = PythonEngineVersion.IronPython2;
+        private PythonEngineVersion engine = PythonEngineVersion.Unspecified;
 
         [JsonConverter(typeof(StringEnumConverter))]
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Populate)]
+        [DefaultValue("IronPython2")]
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModels/PythonNode.cs:39]: can we use nameof?

File: src/Libraries/PythonNodeModels/PythonEngineSelector .cs [hunk 23]
```diff
@@ -9,6 +9,7 @@ namespace PythonNodeModels
     /// </summary>
     public enum PythonEngineVersion
     {
+        Unspecified,
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModels/PythonEngineSelector .cs:12]: this should go at the end of the list as this change re-orders all the existing enums - ie:
Ironpython2 used to = `0`, but after this change it is `1`.

File: src/Libraries/PythonNodeModels/PythonNode.cs [hunk 24]
```diff
@@ -29,15 +32,25 @@ public PythonCodeMigrationEventArgs(string oldCode, string newCode)
...
         /// <summary>
         /// Return the user selected python engine enum.
         /// </summary>
         public PythonEngineVersion Engine
         {
-            get { return engine; }
+            get
+            {
+                // This is a first-time case for newly created nodes only
+                if (engine == PythonEngineVersion.Unspecified)
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModels/PythonNode.cs:48]: 👍 , but please see comment about enum order

File: src/Libraries/PythonNodeModels/PythonNode.cs [hunk 25]
```diff
@@ -48,6 +61,47 @@ public PythonEngineVersion Engine
             }
         }
 
+        private ObservableCollection<PythonEngineVersion> availableEngines;
+        /// <summary>
+        /// Available Python engines.
+        /// </summary>
+        public ObservableCollection<PythonEngineVersion> AvailableEngines
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModels/PythonNode.cs:68]: same question about auto properties

File: src/DynamoCore/Models/DynamoModel.cs (hunk 16 above)
Comment (mmisol) [reply src/DynamoCore/Models/DynamoModel.cs:432]: Can use with private set. Will change

File: test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs [hunk 26]
```diff
@@ -15,7 +15,7 @@ namespace DynamoCoreWpfTests
 {
     public class PythonNodeCustomizationTests : DynamoTestUIBase
     {
-        private readonly List<string> expectedEngineMenuItems = Enum.GetNames(typeof(PythonNodeModels.PythonEngineVersion)).ToList();
+        private readonly List<string> expectedEngineMenuItems = Enum.GetNames(typeof(PythonNodeModels.PythonEngineVersion)).Skip(1).ToList();
```
Comment (mjkkirschner) [inline test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs:18]: ? comment on this maybe?

File: src/DynamoCore/Models/DynamoModel.cs (hunk 17 above)
Comment (mmisol) [reply src/DynamoCore/Models/DynamoModel.cs:572]: The next line is a `TODO` without a prefix, so I'll add it there.

File: src/DynamoCore/Models/DynamoModel.cs (hunk 18 above)
Comment (mmisol) [reply src/DynamoCore/Models/DynamoModel.cs:574]: I don't know in which context this gets serialized/deserialized. XML serialization OOTB does not throw in cases like this but I don't know what you are doing exactly. Can you point to where that happens?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 19 above)
Comment (mmisol) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:73]: I didn't want to initialize this more than once.

File: src/DynamoCoreWpf/ViewModels/Core/SimpleViewModel.cs (hunk 20 above)
Comment (mmisol) [reply src/DynamoCoreWpf/ViewModels/Core/SimpleViewModel.cs:9]: Ok

File: src/DynamoCoreWpf/ViewModels/ViewModelBase.cs (hunk 21 above)
Comment (mmisol) [reply src/DynamoCoreWpf/ViewModels/ViewModelBase.cs:24]: That's a good question. I saw it was not being used. You could say removing it is a breaking change, but it was added a couple of weeks ago, probably by mistake.

File: src/Libraries/PythonNodeModels/PythonNode.cs (hunk 22 above)
Comment (mmisol) [reply src/Libraries/PythonNodeModels/PythonNode.cs:39]: Yes

File: src/Libraries/PythonNodeModels/PythonEngineSelector .cs (hunk 23 above)
Comment (mmisol) [reply src/Libraries/PythonNodeModels/PythonEngineSelector .cs:12]: Technically this only breaks binary compatibility. I put this in the balance with following best practices, and opted for the latter. Let me know if you'd rather change this anyway.

File: src/Libraries/PythonNodeModels/PythonNode.cs (hunk 25 above)
Comment (mmisol) [reply src/Libraries/PythonNodeModels/PythonNode.cs:68]: This is for the purpose of initializing it only once.

File: test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs (hunk 26 above)
Comment (mmisol) [reply test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs:18]: I think I'd rather change the code to make it more clear.

File: src/Libraries/PythonNodeModels/PythonEngineSelector .cs (hunk 23 above)
Comment (mjkkirschner) [reply src/Libraries/PythonNodeModels/PythonEngineSelector .cs:12]: my experience with this type of change is pretty complex confusion - code which was previously compiled against Dynamo 2.7 and selected engine ironPython2 will now be selecting unspecified.

I guess it's unlikely that any clients besides our own would ever be in this situation though.

If you've already thought about it then I'm fine with your decision. 👍

File: src/Libraries/PythonNodeModels/PythonNode.cs (hunk 25 above)
Comment (mjkkirschner) [reply src/Libraries/PythonNodeModels/PythonNode.cs:68]: I guess I'm still a bit confused by this answer - 
can't you do `Public ObserverableCollection<PythonEngineVersion> {get; set;} = ObserverableCollection<PythonEngineVersion>(){ yada yada}` ?

Comment (mmisol) [reply src/Libraries/PythonNodeModels/PythonNode.cs:68]: Yes, you are right. I guess I don't really like to initialize in that way when it takes more than one line. If it was just creating the collection I wouldn't mind.

Comment (mmisol) [reply src/Libraries/PythonNodeModels/PythonNode.cs:68]: Oh, I see you put `{ yada yada}` in same line. I don't really like that either 😄

File: src/DynamoCore/Models/DynamoModel.cs (hunk 17 above)
Comment (QilongTang) [reply src/DynamoCore/Models/DynamoModel.cs:565]: Thanks for adding it @mmisol , can you mark it specially for 3.0? A bit easier to track all the 3.0 issues in code base

Comment (mmisol) [reply src/DynamoCore/Models/DynamoModel.cs:565]: Will do 👍

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 27]
```diff
@@ -583,6 +604,11 @@ public static DynamoViewModel Start(StartConfiguration startConfiguration = new
 
         protected DynamoViewModel(StartConfiguration startConfiguration)
         {
+            // These options are currently fixed. Eventually we should make them dynamic I guess.
+            defaultPythonEngineOptions.Add(new NameValuePairViewModel<string>(string.Empty, WpfResources.DefaultPythonEngineNone));
+            defaultPythonEngineOptions.Add(new NameValuePairViewModel<string>("IronPython2", "IronPython2"));
+            defaultPythonEngineOptions.Add(new NameValuePairViewModel<string>("CPython3", "CPython3"));
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:610]: Sry to get to it late, this is the reason why I mentioned earlier if we want to just use the enums we already exposing as `PythonEngineVersion`, we just need to add a default entry

Comment (QilongTang) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:610]: I guess what I mean is instead of defining a fix collection but loop through the enums and construct it to reduce defined strings. Not sure if you have thought about it already

Comment (mmisol) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:610]: The thing is that while `IronPython2` and `CPython3` are engine names and can go as they are, `Unspecified` is the absence of an engine and should be translated. Probably in different ways according to the context.
Also, not sure how permanent our enum is going to live forever. If we are going to open up the possibility for new engines, we'll need a solution based on names/strings.

**Comment by mmisol** [comment]
@QilongTang If you are ok I'll go ahead and merge this.

**Comment by QilongTang** [comment]
@mmisol I am OK to merge. Since this piece of work will eventually go into D4R and Civil3D. I would like us do some testing and keep it in mind that we may revisit some details here. I like the fact we are using preference setting, just seems more manageable to me. Good work!

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml (hunk 5 above)
Comment (QilongTang) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml:734]: @mmisol Looking back on this, what limitations are in this case? Is it because we want to be dynamic on this and not achievable using child menu items?
